cg/meta/workflow/mutant/quality_controller/quality_controller.py

```diff
@@ -0,0 +1,128 @@
+
+        return sample_results
+
+    def quality_control_sample(
+        self, sample_id: str, quality_metrics: QualityMetrics
+    ) -> SampleQualityResult:
+        sample_metadata = quality_metrics.samples_metadata.samples[sample_id]
+
+        valid_read_count: bool = has_valid_total_reads(sample_metadata)
+
+        if (
+            sample_metadata.is_internal_negative_control
+            or sample_metadata.is_external_negative_control
+        ):
+            sample_quality = SampleQualityResult(
+                sample_id=sample_id,
+                passes_qc=valid_read_count,
+                is_external_negative_control=sample_metadata.is_external_negative_control,
+                is_internal_negative_control=sample_metadata.is_internal_negative_control,
+                passes_reads_threshold=valid_read_count,
+            )
+            return sample_quality
+
+        sample_results = quality_metrics.samples_results.samples[sample_id]
+
+        sample_passes_qc: bool = valid_read_count and sample_results.qc_pass
+
+        sample_quality = SampleQualityResult(
+            sample_id=sample_id,
+            passes_qc=sample_passes_qc,
+            is_external_negative_control=sample_metadata.is_external_negative_control,
+            is_internal_negative_control=sample_metadata.is_internal_negative_control,
+            passes_reads_threshold=valid_read_count,
+            passes_mutant_qc=sample_results.qc_pass,
+        )
+
+        ResultLogger.log_sample_result(sample_quality)
+        return sample_quality
+
+    def quality_control_case(self, sample_results: list[SampleQualityResult]) -> CaseQualityResult:
+        internal_negative_control_pass_qc: bool = internal_negative_control_qc_pass(sample_results)
+        external_negative_control_pass_qc: bool = external_negative_control_qc_pass(sample_results)
+
+        case_pass_qc: bool = self.case_qc_pass(sample_results=sample_results)
+
+        result = CaseQualityResult(
+            passes_qc=(
+                case_pass_qc
+                and internal_negative_control_pass_qc
+                and external_negative_control_pass_qc
+            ),
+            internal_negative_control_passes_qc=internal_negative_control_pass_qc,
+            external_negative_control_passes_qc=external_negative_control_pass_qc,
+        )
+        ResultLogger.log_case_result(result)
+        return result
+
+    def case_qc_pass(self, sample_results: list[SampleQualityResult]) -> bool:
+        total_samples: int = 0
+        failed_samples: int = 0
+
+        for sample_result in sample_results:
+            if (
+                not sample_result.is_external_negative_control
+                and not sample_result.is_internal_negative_control
+            ):
+                total_samples += 1
+                if not sample_result.passes_qc:
+                    failed_samples += 1
+
+        return failed_samples / total_samples < MutantQC.FRACTION_OF_SAMPLES_WITH_FAILED_QC_TRESHOLD
```

Comment:
Could you rewrite this part to make it easier to read/understand?

```
def case_qc_pass(self, sample_results: list[SampleQualityResult]) -> bool:
    valid_samples = [
        sample_result for sample_result in sample_results 
        if not sample_result.is_external_negative_control 
        and not sample_result.is_internal_negative_control
    ]
    failed_samples = sum(not sample_result.passes_qc for sample_result in valid_samples)
    return failed_samples / len(valid_samples) < MutantQC.FRACTION_OF_SAMPLES_WITH_FAILED_QC_TRESHOLD if valid_samples else False
```

Reply:
Yes! Thanks for the suggestion!